### PR TITLE
add typing  and py.typed

### PR DESCRIPTION
**Pull request by trim21**
consider we still support python2.7, this PR add type stub instead of inline types.

### CODE REVIEW

**Comment by trim21** [comment]
@jquast

**Comment by GalaxySnail** [comment]
Sorry to keep you waiting. I'm currently working on running tests using GitHub Actions, and the next step will be integrating mypy.

**Comment by trim21** [comment]
> Sorry to keep you waiting. I'm currently working on running tests using GitHub Actions, and the next step will be integrating mypy.

I can help with that

**Comment by trim21** [comment]
@GalaxySnail

**Comment by trim21** [comment]
@GalaxySnail @jquast can this PR get some code review？

**Comment by GalaxySnail** [comment]
I'm still trying to figure out what the best practice is to test typing stubs, especially back in Python 2.7. Do you have any ideas?

**Comment by trim21** [comment]
> I'm still trying to figure out what the best practice is to test typing stubs, especially back in Python 2.7. Do you have any ideas?

run mypy stubtest to check if type stub is in-sync with source code I guess? Without inline type annotations there is no need to run mypy I think.

https://mypy.readthedocs.io/en/stable/stubtest.html

No need to run stubtest with runtime 2.7, you can run it in latest python version. type stub are expected to be parsed as latest python version.

**Comment by trim21** [comment]
> I'm still trying to figure out what the best practice is to test typing stubs, especially back in Python 2.7. Do you have any ideas?

just add ci job for type stubs testing

**Comment by trim21** [comment]
@GalaxySnail any review?

**Comment by jquast** [comment]
Looks OK to me... @GalaxySnail ?

**Review by jquast — APPROVED** [review]

**Comment by trim21** [comment]
CI is broken

**Comment by GalaxySnail** [comment]
> CI is broken

Yes. (That's what tests intend to do.) We should fix them before merging.

**Comment by trim21** [comment]
> > CI is broken
> 
> Yes. (That's what tests intend to do.) We should fix them before merging.

no... As I mentioned before, you only need to run mypy's stubtest in latest python, no need to run it in old python version.

File: wcwidth/wcwidth.pyi [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from typing import Optional
+
+
+def wcwidth(wc: str, unicode_version: str = ...) -> int:
+    ...
+
+
+def wcswidth(pwcs: str, n: Optional[int] = None, unicode_version: str = ...):
+    ...
```
Comment (GalaxySnail) [inline wcwidth/wcwidth.pyi:9]: These 2 functions accept [`typing.AnyStr`](https://docs.python.org/3/library/typing.html#typing.AnyStr) instead of `str`.

**Comment by GalaxySnail** [comment]
> no... As I mentioned before, you only need to run mypy's stubtest in latest python, no need to run it in old python version.

I doubt it. If someone is using wcwidth on an older python version, they will fail just like what happenned in CI, which makes typing stubs useless.

**Comment by GalaxySnail** [comment]
> and ci failed because you try to run mypy in old version of python, mypy need python>=3.6. it has nothing to do with type stub.

OK, thanks for your clarification. So we should run mypy on python>=3.6 only. AFAIK the latest version of mypy which supports python2 is mypy 0.971 [1].

[1] https://mypy-lang.blogspot.com/2022/07/mypy-0971-released.html

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -167,3 +167,16 @@ jobs:
...
+  typing:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.x'
+      - run: pip install mypy
+      - run: pip install -e .
```
Comment (GalaxySnail) [inline .github/workflows/ci.yml:180]: It's better to call tox here.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -167,3 +167,16 @@ jobs:
...
+      - name: Run tests
+        run: stubtest wcwidth
```
Comment (GalaxySnail) [inline .github/workflows/ci.yml:182]: A missing EOL.

File: wcwidth/wcwidth.pyi (hunk 1 above)
Comment (GalaxySnail) [reply wcwidth/wcwidth.pyi:9]: Hmmmmmm... I was wrong, they don't accept `bytes` on Python 3, but accept both `str` and `unicode` on Python 2.

**Comment by trim21** [comment]
> > and ci failed because you try to run mypy in old version of python, mypy need python>=3.6. it has nothing to do with type stub.
> 
> OK, thanks for your clarification. So we should run mypy on python>=3.6 only. AFAIK the latest version of mypy which supports python2 is mypy 0.971 [1].
> 
> [1] https://mypy-lang.blogspot.com/2022/07/mypy-0971-released.html

we don't need to use this version， only downstream developers who are still using Python2 need to use this version， we can run mypy in python 3.12 and use latest mypy version.

File: wcwidth/wcwidth.pyi (hunk 1 above)
Comment (trim21) [reply wcwidth/wcwidth.pyi:9]: Thanks, I'll fix this later

**Comment by GalaxySnail** [comment]
> > OK, thanks for your clarification. So we should run mypy on python>=3.6 only. AFAIK the latest version of mypy which supports python2 is mypy 0.971 [1].
> > [1] https://mypy-lang.blogspot.com/2022/07/mypy-0971-released.html
> 
> we don't need to use this version， only downstream developers who are still using Python2 need to use this version， we can run mypy in python 3.12 and use latest mypy version.

No, if we want downstream users to use it, we should test it, otherwise it will probably break for downstream users.

**Comment by trim21** [comment]
> > > OK, thanks for your clarification. So we should run mypy on python>=3.6 only. AFAIK the latest version of mypy which supports python2 is mypy 0.971 [1].
> > > [1] https://mypy-lang.blogspot.com/2022/07/mypy-0971-released.html
> > 
> > 
> > we don't need to use this version， only downstream developers who are still using Python2 need to use this version， we can run mypy in python 3.12 and use latest mypy version.
> 
> No, if we want downstream users to use it, we should test it, otherwise it will probably break for downstream users.

For python code, yes, I agree with you. But typing stub is another scope.

**Comment by trim21** [comment]
> > > > OK, thanks for your clarification. So we should run mypy on python>=3.6 only. AFAIK the latest version of mypy which supports python2 is mypy 0.971 [1].
> > > > [1] https://mypy-lang.blogspot.com/2022/07/mypy-0971-released.html
> > > 
> > > 
> > > we don't need to use this version， only downstream developers who are still using Python2 need to use this version， we can run mypy in python 3.12 and use latest mypy version.
> > 
> > 
> > No, if we want downstream users to use it, we should test it, otherwise it will probably break for downstream users.
> 
> For python code, yes, I agree with you. But typing stub is another scope, they are not actually code to be executed.

File: wcwidth/wcwidth.pyi (hunk 1 above)
Comment (trim21) [reply wcwidth/wcwidth.pyi:9]: So looks like there is no way to type this correctly?

**Comment by trim21** [comment]
Sorry, the test you added failed ci, I don't know how to fix it.

**Comment by jquast** [comment]
Thank you both all the attention and efforts and concerns. I understand that python 2 compatibility is difficult, especially for typing.

I have created Issue #102 with my thoughts about dropping support. I think restricting future releases to python 3.6+ would allow easier typing in-file, without stubs and make this easier for everyone?

**Comment by trim21** [comment]
> Thank you both all the attention and efforts and concerns. I understand that python 2 compatibility is difficult, especially for typing.
> 
> I have created Issue #102 with my thoughts about dropping support. I think restricting future releases to python 3.6+ would allow easier typing in-file, without stubs and make this easier for everyone?

I can wait until then

**Pull request closed without merging**